### ISSUE TITLE
CD workflow for fabric8 apps

### ISSUE THREAD

**Issue by rawlingsj** [fabric8io/fabric8#4437]
So thinking about creating a docker workflow to automate the build, test and release of one of our quickstart apps.  I figured I'd write out the steps I'm thinking then we can all pitch in with ideas or thoughts on how it should work.  Once done we can turn this into docs and record a demo.

OK so lets use the hubot slack app to start with.  It's simple, small, not in use by anyone yet to my knowledge and doesn't depend on any other apps running.

There is a hubot-slack docker file here https://github.com/fabric8io/hubot-slack that currently has an automated build in Dockerhub setup to build and tag an image as :latest on every commit.

This is useful during development but not practical in a CI / CD env when we need to release an image using docker tags as versions. During a release, triggered by a human interaction should use the exact image that has been tested and apply a new tag which is an incremented version number.  Once this has completed an update is needed to the hubot-slack fabric8 app (currently living https://github.com/fabric8io/quickstarts/tree/master/apps/hubot-slack) to reference the new docker tag / version.  
Once this step is complete the app should be automatically tested and released to maven central.  For this reason we should split out the app from the quickstarts repo so it can be versioned / released independently from the others.

To do this we will use the docker workflow plugin to chain steps together, fan out integration tests (probably only one for this app), fan in, push to docker hub, notify irc / letschat or slack that a latest image  is available in dockerhub.  Await / offer human trigger to run release job to maven central and notify once complete.

Questions:
- Does this sounds like a decent high level start?
- Human trigger for release to be via jenkins job / docker workflow / hubot?   
- Should we have a catalogue or naming standard to follow for apps so they can be filtered searched?
- Can we reference the app icon in each readme and also be automatically referenced / displayed on the fabric8 website?

**Comment by jimmidyson** [comment]
Each app in it's own git repo I assume? Versioned independently?

**Comment by rawlingsj** [comment]
@jimmidyson yeah correct each app should be in it's own git repo

**Comment by rawlingsj** [comment]
What would be good is for the human approval step to be via jenkins but you could also trigger it using hubot, something like..
`> fabric8 release hubot-slack-app v1.0.1`
Hubot would then trigger the Jenkins docker workflow.

**Comment by jstrachan** [comment]
Right now there are so many manual steps to getting a release out; I was hoping we'd automate the following flow ASAP first:

* release fabric8io/fabric8
* submitting the PR on fabric8io/quickstarts with the new version of `<fabric8.version>...</fabric8.version` in each pom.xml
* when the PR is merged release fabric8io/quickstarts with the same version
* when thats done and a human approvaes submit a PR on fabric8io/fabric8 to update the website (replacing the version in `fabric8io/**/*.md` and `fabric8io/website/src/**.*` excluding `Changes.md`).

Once that works well, I figured we'd split up the huge quickstarts repo into smaller pieces. I was initially thinking 1 repo for devops and one for ipaas then separate git repos for each quickstart. (Then users who want to work on a specific quickstart can just clone a tiny repo).

In terms of, say, a specific chat app; we've a mvn project to create the k8s json and something to make a docker image. We could merge those into the same git repo. So that as we make a new tagged docker version, we then update the pom.xml and do a mvn release of the new k8s json. i.e. automate that for every tagged docker image we always generate a k8s too.

Then the courser grained projects like 'devops' and 'ipaas' need to be updated to refer to this new version (as we use mvn dependencies to combine templates together for things like devops / ipaas).

Though TBH - we could just leave the k8s generation in the courser grained projects like devops/ipaas; it might be a little simpler then. i.e. we have small git repos that just make docker images; then a mvn build in devops/ipaas that makes the k8s json and aggregates it into courser grained templates?

**Comment by rawlingsj** [comment]
Yeah fair enough, the other reason I was suggesting starting with the apps was because a few of them dont have tags in dockerhub yet and I lost an hour to that yesterday.  Having said that you're point about the dockerfile and the k8s config being collocated feels better so yeah lets get fabric8 done first and then tackle the apps after.  It would be good once we're there though to tackle the docker images that dont have a release number as a tag first.

Currently the search and replace utility image isn't requesting a PR, instead it just pushes directly back to the repo it cloned.  I'll update that then we should be ready for the steps above to turn into the workflow.
